include/swift/AST/Attr.h

```diff
@@ -2140,6 +2140,30 @@ class TransposeAttr final
   }
 };
 
+/// The `@hasAsyncAlternative` attribute marks a function as having an async
+/// alternative, optionally providing a name (for cases when the alternative
+/// has a different name).
+class HasAsyncAlternativeAttr final : public DeclAttribute  {
+public:
+  /// An optional name of the async alternative function, where the name of the
+  /// attributed function is used otherwise.
+  const DeclNameRef Name;
```

Comment:
We'll want type checking to resolve this to a Decl*, and serialization to serialize the Decl reference rather than the name (so clients don't have to resolve the Decl again).